n3fit/src/n3fit/layers/DIS.py

```diff
@@ -41,39 +45,19 @@ def gen_mask(self, basis):
                 basis_mask[i] = True
         return op.numpy_to_tensor(basis_mask, dtype=bool)
 
-    def call(self, pdf):
-        """
-        This function perform the fktable \otimes pdf convolution.
+    def mask_fk(self, fk, mask):
+        return op.einsum('fF, nFx -> nxf', mask, fk)
 
-        First pass the input PDF through a mask to remove the unactive flavors,
-        then a tensor_product between the PDF and each fktable is performed
-        finally the defined operation is applied to all the results
+    def build(self, input_shape):
+        super().build(input_shape)
+        if self.num_replicas > 1:
 
-        Parameters
-        ----------
-            pdf:  backend tensor
-                rank 4 tensor (batch_size, replicas, xgrid, flavours)
+            def compute_observable(pdf, masked_fk):
+                return op.einsum('brxf, nxf -> brn', pdf, masked_fk)
```

Comment:
I don't understand these internal functions.

As far as I can see it is not taking anything from the outside scope so you could make a few of them module-level functions, say

`compute_observable_multireplica`
`compute_observable_one_replica`

etc, with a docstr explaining the rationale

and then according to the if conditional do `self.compute_observable = compute_observable`

Reply:
I agree that looks better, done!